manager/integration/tests/test_scheduling.py

```diff
@@ -454,6 +468,424 @@ def test_data_locality_basic():
+
+    replica_node_soft_anti_affinity_setting = \
+        client.by_id_setting(SETTING_REPLICA_NODE_SOFT_ANTI_AFFINITY)
+    try:
+        client.update(replica_node_soft_anti_affinity_setting,
+                      value="false")
+    except Exception as e:
+        print("Exception when update "
+              "Replica Node Level Soft Anti-Affinity setting",
+              replica_node_soft_anti_affinity_setting, e)
+
+    volume3_name = volume_name + "-3"
+    volume3_size = str(1 * Gi)
+    volume3_data_path = "/data/test"
+    pv3_name = volume3_name + "-pv"
+    pvc3_name = volume3_name + "-pvc"
+    pod3_name = volume3_name + "-pod"
+    pod3 = pod
+
+    pod3['metadata']['name'] = pod3_name
+
+    volume3 = client.create_volume(name=volume3_name,
+                                   size=volume3_size,
+                                   numberOfReplicas=1)
+
+    volume3 = wait_for_volume_detached(client, volume3_name)
+    volume3 = client.by_id_volume(volume3_name)
+    create_pv_for_volume(client, core_api, volume3, pv3_name)
+    create_pvc_for_volume(client, core_api, volume3, pvc3_name)
+
+    volume3 = client.by_id_volume(volume3_name)
+
+    pod3['spec']['volumes'] = [{
+        "name": "pod-data",
+        "persistentVolumeClaim": {
+            "claimName": pvc3_name
+        }
+    }]
+
+    pod3['spec']['nodeSelector'] = {"kubernetes.io/hostname": node3.name}
+    create_and_wait_pod(core_api, pod3)
+    volume3 = wait_for_volume_healthy(client, volume3_name)
+
+    write_pod_volume_random_data(core_api, pod3_name,
+                                 volume3_data_path, DATA_SIZE_IN_MB_4)
+
+    volume3.updateDataLocality(dataLocality="best-effort")
+    volume3 = client.by_id_volume(volume3_name)
+
+    if volume3.replicas[0]['hostId'] != node3.name:
+        wait_for_rebuild_start(client, volume3_name)
+        volume3 = client.by_id_volume(volume3_name)
+        assert len(volume3.replicas) == 2
+        wait_for_rebuild_complete(client, volume3_name)
+
+    volume3 = wait_for_volume_replica_count(client, volume3_name, 1)
+    assert volume3.replicas[0]["hostId"] == node3.name
+
+    delete_and_wait_pod(core_api, pod3_name)
+
+    pod3['spec']['nodeSelector'] = {"kubernetes.io/hostname": node1.name}
+    create_and_wait_pod(core_api, pod3)
+
+    wait_for_rebuild_start(client, volume3_name)
+    volume3 = client.by_id_volume(volume3_name)
+    volume3.detach()
+    volume3 = client.by_id_volume(volume3_name)
+    assert len(volume3.replicas) == 2
+    volume3 = client.by_id_volume(volume3_name)
+
+    err_replica = None
+    for replica in volume3.replicas:
+        if replica["hostId"] == node1.name:
+            err_replica = replica
+            break
+
+    assert err_replica is not None
+    wait_for_replica_failed(client, volume3_name, err_replica['name'])
+
+    delete_and_wait_pod(core_api, pod3_name)
+    volume3 = wait_for_volume_detached(client, volume3_name)
```

Comment:
Test still fails on case 3 here

Error: 
```
MountVolume.SetUp failed for volume "longhorn-testvol-skp0d3-3-pv" : rpc error: code = InvalidArgument desc = There is no block device frontend for volume longhorn-testvol-skp0d3-3
```

Reply:
hmm. Is it because we detach the volume manually at https://github.com/longhorn/longhorn-tests/blob/c66584c2b280967cb3dba7c54e6fab9e3b349990/manager/integration/tests/test_scheduling.py#L740
so the engine of the volume is killed at that moment and the pod can not find the block device frontend (which is the engine)?
If it is the case, can we try this?
```
    wait_for_rebuild_start(client, volume3_name)
     volume3 = client.by_id_volume(volume3_name)
     delete_and_wait_pod(core_api, pod3_name)
     volume3 = client.by_id_volume(volume3_name)
     assert len(volume3.replicas) == 2
     volume3 = client.by_id_volume(volume3_name)

      err_replica = None
     for replica in volume3.replicas:
         if replica["hostId"] == node1.name:
             err_replica = replica
             break

      assert err_replica is not None
     wait_for_replica_failed(client, volume3_name, err_replica['name'])

  
     volume3 = wait_for_volume_detached(client, volume3_name)
```

Reply:
@PhanLe1010 
Time waiting for pod deletion may exceed replica rebuilding time, causing the test to fail
I've updated the test to work around the issue by killing volume engine to simulate rebuild interruption